Optional constructor arguments injection, private properties injection

Rastusik opened the pull request:
Hi,

I needed to update your package for the project I'm currently working on in Pexel Federation. I added the following two features:
- Optional constructor arguments injection - the ability to configure the selected constructor arguments (even if they are the last in the arguments list), all the other constructor arguments are wutowired by default.
- Private properties injection - it is possible to inject not only public properties, but also private and protected ones

What dou you think about these features? Is it possible to merge them into your package, or should I create my own fork of the package?

Commit:
ability to typehint just selected parameters (interfaces) using services.yml, ability to inject private properties

new features documentation, code cleanup, comments

README.md cosmetic fix (code snipplet language added)

some code smell fixes

added optional constructor dependency test

gitignore and composer.json cleanup - removed phpunit dependency

removed private property injection feature

Method injection fix

jakubkulhan commented:
Hello, thank you for the PR! 

Optional constructor arguments injection is a superb addition to autowiring capabilities. If you add test to [AutowiringCompilerPassTest](https://github.com/skrz/autowiring-bundle/blob/master/Tests/DependencyInjection/Compiler/AutowiringCompilerPassTest.php), I'll be happy to merge this feature.

About injecting private/protected properties, you do not want to go this path... It doesn't improve encapsulation and at the same time it sacrifices runtime perfomance. Symfony container's `PhpDumper` doesn't support this for a reason.

Rastusik commented:
Ok, I'll provide the test.

Regarding the private properties injection - I don't think there would be a measurable decrease of performance. Anyway, the problem isn't about me wanting to go this way, I just want it for one special use case (where I think it would come handy) - I don't want to to have my controllers dependent on the symfony di container and at the same time I don't want to inject all the needed properties using the constructor. Using public methods seems dirty to me, since it's weird to have e.g. the token storage, or request stack publicly accessible from the controller.

jakubkulhan commented:
> I don't think there would be a measurable decrease of performance.

https://3v4l.org/EmrUe

It can take up to 10ms to inject private properties using `ReflectionProperty` on mid-size container (500 services, about 10 deps every service). Compared to setting public property which is sub-millisecond.

> since it's weird to have e.g. the token storage, or request stack publicly accessible from the controller.

It's much worse to hide that the controller has the deps and not to be able to create a functioning instance of the controller without the DI container (or using reflection again).

Rastusik commented:
I pushed the test for the optional constructor injection.

> It can take up to 10ms to inject private properties using ReflectionProperty on mid-size container (500 services, about 10 deps every service). Compared to setting public property which is sub-millisecond.

Yes, but in some verions of php 5.6 and in php 7 it is only 0.007 seconds. And for my special case, I just need 6 dependencies, not 5000. I don't prefer to inject private methods in general, I just like the possibility to be able to do it on occasion. Othe DI frameworks allow this, e.g. PHP-DI or http://jmsyst.com/bundles/JMSDiExtraBundle and it is also possoble to do that in Java DI frameworks. I still know, that it isn't best practise, I just would like to have this possibility.

If you still think this is a no-go, I can delete the code from the pull request, but I'd prefer to let it there anyway :)

jakubkulhan commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 /vendor
 composer.lock
+.idea/*
```
I suggest (as [I was suggested before](https://github.com/skrz/autowiring-bundle/pull/2/files#r35136818)) to use `~/.gitignore_global` to ignore IDE's projects files. This is not necessary.

jakubkulhan commented on composer.json:
```diff
@@ -15,6 +15,9 @@
         "symfony/http-kernel": "~2.7"
     },
+    "require-dev": {
+        "phpunit/phpunit": "5.0.*"
```
The bundle has to maintain PHP 5.4 compatibility, PHPUnit 5.0.\* fails build because it supports php >= 5.6 https://travis-ci.org/skrz/autowiring-bundle/builds/87307234

It is better not to have a library depending on phpunit or other tools. Please don't add this.

jakubkulhan commented:
Even JMSDiExtraBundle [states](http://jmsyst.com/bundles/JMSDiExtraBundle/master/usage#property-injection): _Injecting into private, or protected properties is only supported on controllers,_ because in order to do so it has to generate special factory class for such cases. 

You might want to send PR to the Symfony DI component with altered `PhpDumper` to support private property injection, supplying its own `PhpDumper` is not a goal of this bundle. Sorry, private property injection won't get merged :)

Please leave out private property injection and then squash the commits.

Rastusik commented:
I removed the private property injection feature. If the feature was enabled only for the controllers like in JMSDiExtraBundle - would that still be a problem? :D

jakubkulhan commented:
> If the feature was enabled only for the controllers like in JMSDiExtraBundle - would that still be a problem? :D

That something can be done doesn't mean it should be done :)

Ok, the feature is ready to merge. The last thing, please [squash your commits into one](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/). Thank you.

Rastusik commented:
I squashed the commit as you requested.

Please check one more thing - I think I fixed a bug in the method injection feature https://github.com/skrz/autowiring-bundle/commit/27aadf693e7dce34e51694fcc19bf8ac3561ff91#diff-22d46138c9deebd18edad9eae1211916L172

The autowiring function got constructor arguments as arguments for the method injection, which I think is wrong (at least I got errors without my fix)

jakubkulhan commented:
Thank you!

> Please check one more thing - I think I fixed a bug in the method injection feature 27aadf6#diff-22d46138c9deebd18edad9eae1211916L172

Yes, this was a bug.
